packages/react-dom/src/client/ReactDOMFiberComponent.js

```diff
@@ -927,14 +946,28 @@ export function diffHydratedProperties(
       if (typeof nextProp === 'string') {
         if (domElement.textContent !== nextProp) {
           if (__DEV__ && !suppressHydrationWarning) {
-            warnForTextDifference(domElement.textContent, nextProp);
+            // Determine whether renderToString() was used to produce the HTML contents
+            // of the root element by checking if the "data-reactroot" attribute exists
+            // in the root element.
+            if (!domElement.hasAttribute('data-reactroot')) {
```

Comment:
I don't quite understand how this check would work.

`domElement` here can be arbitrarily nested, can it not? Therefore it's not guaranteed to be a root.

Reply:
Hmmm... You're right. According to the docs,  if the HTML contents of the element that is passed to `hydrate()` is rendered by `renderToString()`, a `data-reactroot` attribute is created automatically, is it not?

Reply:
Doc: https://reactjs.org/docs/react-dom-server.html#rendertostaticmarkup

Reply:
@cramaechi you're right that a `data-reactroot` attribute would be added, but it's only added to the _root_ element. So if you render a component like:

```jsx
const App = () => (
  <div>
   <h1>Hello</h1>
  </div>
)
```

with `renderToString` you'll get markup that looks like:

```html
<div data-reactroot>
 <h1>Hello, world</h1>
</div>
```

You'll see that the `data-reactroot` attribute only exists on the root element (the `div`) but not any of it's children (the `h1` tag). The problem with with the approach you took here is that it will be potentially called with elements that aren't the root, so it will end up warning when it shouldn't. In this case, it might see the `h1` tag doesn't have the `data-reactroot` attribute, and warn, even though it shouldn't.

Does that make sense? To implement this correctly you'll need to add a check that _only_ runs on the root element passed to `ReactDOM.hydrate`

Reply:
Hi @aweary - That makes perfect sense.

And thank you for the much needed, detailed explanation!

I will go ahead and make the changes.

Reply:
@aweary - In detecting the root element, would checking the node type of the element work?

I ask this because if `div` is a root element in your example, would it actually qualify as a document node?

Reply:
@cramaechi see my comment here for my recommendation: https://github.com/facebook/react/pull/12166#issuecomment-363562627

Reply:
Hi @aweary- I think the hyperlink is broken.